### DISCUSSION TITLE
Available probability distribution in Aesara

### DISCUSSION THREAD

**Discussion by rlouf** [aesara-devs/aesara#1093] [General]
I gathered the following table which enumerates the distributions available in `numpy` or `aesara.tensor.random.basic` as well as whether they can be transpiled to Numba/JAX:

<table border="2" cellspacing="0" cellpadding="6" rules="groups" frame="hsides">

<colgroup>
<col  class="org-left" />

<col  class="org-left" />

<col  class="org-left" />

<col  class="org-left" />

<col  class="org-left" />
</colgroup>
<thead>
<tr>
<th scope="col" class="org-left">distribution</th>
<th scope="col" class="org-left">Numpy</th>
<th scope="col" class="org-left">Aesara</th>
<th scope="col" class="org-left">Aesara (Numba)</th>
<th scope="col" class="org-left">Aesara (JAX)</th>
</tr>
</thead>

<tbody>
<tr>
<td class="org-left">integers</td>
<td class="org-left">x</td>
<td class="org-left">x</td>
<td class="org-left">&#xa0;</td>
<td class="org-left">#1331</td>
</tr>

<tr>
<td class="org-left">random</td>
<td class="org-left">x</td>
<td class="org-left">x</td>
<td class="org-left">&#xa0;</td>
<td class="org-left">&#xa0;</td>
</tr>

<tr>
<td class="org-left">choice</td>
<td class="org-left">x</td>
<td class="org-left">x</td>
<td class="org-left">x</td>
<td class="org-left">x</td>
</tr>

<tr>
<td class="org-left">bytes</td>
<td class="org-left">x</td>
<td class="org-left">-</td>
<td class="org-left">-</td>
<td class="org-left">-</td>
</tr>

<tr>
<td class="org-left">betabinomial</td>
<td class="org-left">&#xa0;</td>
<td class="org-left">x</td>
<td class="org-left">&#xa0;</td>
<td class="org-left">#1332</td>
</tr>

<tr>
<td class="org-left">shuffle</td>
<td class="org-left">x</td>
<td class="org-left">#1367</td>
<td class="org-left">&#xa0;</td>
<td class="org-left">x</td>
</tr>

<tr>
<td class="org-left">permutation</td>
<td class="org-left">x</td>
<td class="org-left">x</td>
<td class="org-left">x</td>
<td class="org-left">x</td>
</tr>

<tr>
<td class="org-left">permuted</td>
<td class="org-left">x</td>
<td class="org-left">#1354</td>
<td class="org-left">&#xa0;</td>
<td class="org-left">&#xa0;</td>
</tr>

<tr>
<td class="org-left">bernoulli</td>
<td class="org-left">&#xa0;</td>
<td class="org-left">x</td>
<td class="org-left">x</td>
<td class="org-left">x</td>
</tr>

<tr>
<td class="org-left">beta</td>
<td class="org-left">x</td>
<td class="org-left">x</td>
<td class="org-left">x</td>
<td class="org-left">x</td>
</tr>

<tr>
<td class="org-left">binomial</td>
<td class="org-left">x</td>
<td class="org-left">x</td>
<td class="org-left">x</td>
<td class="org-left">#1321</td>
</tr>

<tr>
<td class="org-left">categorical</td>
<td class="org-left">&#xa0;</td>
<td class="org-left">x</td>
<td class="org-left">x</td>
<td class="org-left">x</td>
</tr>

<tr>
<td class="org-left">cauchy</td>
<td class="org-left">&#xa0;</td>
<td class="org-left">x</td>
<td class="org-left">x</td>
<td class="org-left">x</td>
</tr>

<tr>
<td class="org-left">chisquare</td>
<td class="org-left">x</td>
<td class="org-left">x</td>
<td class="org-left">x</td>
<td class="org-left">#1363</td>
</tr>

<tr>
<td class="org-left">dirichlet</td>
<td class="org-left">x</td>
<td class="org-left">x</td>
<td class="org-left">x</td>
<td class="org-left">x</td>
</tr>

<tr>
<td class="org-left">exponential</td>
<td class="org-left">x</td>
<td class="org-left">x</td>
<td class="org-left">x</td>
<td class="org-left">x</td>
</tr>

<tr>
<td class="org-left">f</td>
<td class="org-left">x</td>
<td class="org-left">#1366</td>
<td class="org-left">&#xa0;</td>
<td class="org-left">&#xa0;</td>
</tr>

<tr>
<td class="org-left">gamma</td>
<td class="org-left">x</td>
<td class="org-left">x</td>
<td class="org-left">x</td>
<td class="org-left">x</td>
</tr>

<tr>
<td class="org-left">gengamma</td>
<td class="org-left">&#xa0;</td>
<td class="org-left">x</td>
<td class="org-left">&#xa0;</td>
<td class="org-left">#1333</td>
</tr>

<tr>
<td class="org-left">geometric (geom)</td>
<td class="org-left">x</td>
<td class="org-left">x</td>
<td class="org-left">x</td>
<td class="org-left">x</td>
</tr>

<tr>
<td class="org-left">gumbel</td>
<td class="org-left">x</td>
<td class="org-left">x</td>
<td class="org-left">x</td>
<td class="org-left">x</td>
</tr>

<tr>
<td class="org-left">halfcauchy</td>
<td class="org-left">&#xa0;</td>
<td class="org-left">x</td>
<td class="org-left">&#xa0;</td>
<td class="org-left">x</td>
</tr>

<tr>
<td class="org-left">halfnormal</td>
<td class="org-left">&#xa0;</td>
<td class="org-left">x</td>
<td class="org-left">x</td>
<td class="org-left">x</td>
</tr>

<tr>
<td class="org-left">hypergeometric (hypergeom)</td>
<td class="org-left">x</td>
<td class="org-left">x</td>
<td class="org-left">x</td>
<td class="org-left">#1324</td>
</tr>

<tr>
<td class="org-left">invgamma</td>
<td class="org-left">&#xa0;</td>
<td class="org-left">x</td>
<td class="org-left">&#xa0;</td>
<td class="org-left">#1368</td>
</tr>

<tr>
<td class="org-left">laplace</td>
<td class="org-left">x</td>
<td class="org-left">x</td>
<td class="org-left">x</td>
<td class="org-left">x</td>
</tr>

<tr>
<td class="org-left">logistic</td>
<td class="org-left">x</td>
<td class="org-left">x</td>
<td class="org-left">x</td>
<td class="org-left">x</td>
</tr>

<tr>
<td class="org-left">lognormal</td>
<td class="org-left">x</td>
<td class="org-left">x</td>
<td class="org-left">x</td>
<td class="org-left">x</td>
</tr>

<tr>
<td class="org-left">logseries (logser)</td>
<td class="org-left">x</td>
<td class="org-left">#1349</td>
<td class="org-left">&#xa0;</td>
<td class="org-left">&#xa0;</td>
</tr>

<tr>
<td class="org-left">multinomial</td>
<td class="org-left">x</td>
<td class="org-left">x</td>
<td class="org-left">x</td>
<td class="org-left">#1326</td>
</tr>

<tr>
<td class="org-left">multivariate_hypergeometric</sub></td>
<td class="org-left">x</td>
<td class="org-left">#1355</td>
<td class="org-left">&#xa0;</td>
<td class="org-left">&#xa0;</td>
</tr>

<tr>
<td class="org-left">multivariate_normal</sub></td>
<td class="org-left">x</td>
<td class="org-left">x</td>
<td class="org-left">&#xa0;</td>
<td class="org-left">x</td>
</tr>

<tr>
<td class="org-left">negative_binomial</sub> (nbinom)</td>
<td class="org-left">x</td>
<td class="org-left">x</td>
<td class="org-left">x</td>
<td class="org-left">#1327</td>
</tr>

<tr>
<td class="org-left">noncentral_chisquare</sub></td>
<td class="org-left">x</td>
<td class="org-left">#1356</td>
<td class="org-left">&#xa0;</td>
<td class="org-left">&#xa0;</td>
</tr>

<tr>
<td class="org-left">noncentral_f</sub></td>
<td class="org-left">x</td>
<td class="org-left">#1357</td>
<td class="org-left">&#xa0;</td>
<td class="org-left">&#xa0;</td>
</tr>

<tr>
<td class="org-left">normal</td>
<td class="org-left">x</td>
<td class="org-left">x</td>
<td class="org-left">x</td>
<td class="org-left">x</td>
</tr>

<tr>
<td class="org-left">pareto</td>
<td class="org-left">x</td>
<td class="org-left">x</td>
<td class="org-left">x</td>
<td class="org-left">x</td>
</tr>

<tr>
<td class="org-left">poisson</td>
<td class="org-left">x</td>
<td class="org-left">x</td>
<td class="org-left">x</td>
<td class="org-left">x</td>
</tr>

<tr>
<td class="org-left">power</td>
<td class="org-left">x</td>
<td class="org-left">#1352</td>
<td class="org-left">&#xa0;</td>
<td class="org-left">&#xa0;</td>
</tr>

<tr>
<td class="org-left">rayleigh</td>
<td class="org-left">x</td>
<td class="org-left">x</td>
<td class="org-left">&#xa0;</td>
<td class="org-left">&#xa0;</td>
</tr>

<tr>
<td class="org-left">randint</td>
<td class="org-left">&#xa0;</td>
<td class="org-left">x</td>
<td class="org-left">x</td>
<td class="org-left">x</td>
</tr>

<tr>
<td class="org-left">standard_cauchy</sub></td>
<td class="org-left">x</td>
<td class="org-left">#1358</td>
<td class="org-left">&#xa0;</td>
<td class="org-left">&#xa0;</td>
</tr>

<tr>
<td class="org-left">standard_exponential</sub></td>
<td class="org-left">x</td>
<td class="org-left">#1358</td>
<td class="org-left">&#xa0;</td>
<td class="org-left">&#xa0;</td>
</tr>

<tr>
<td class="org-left">standard_gamma</sub></td>
<td class="org-left">x</td>
<td class="org-left">#1358</td>
<td class="org-left">&#xa0;</td>
<td class="org-left">&#xa0;</td>
</tr>

<tr>
<td class="org-left">standard_normal</sub></td>
<td class="org-left">x</td>
<td class="org-left">x</td>
<td class="org-left">&#xa0;</td>
<td class="org-left">x</td>
</tr>

<tr>
<td class="org-left">standard_t</sub></td>
<td class="org-left">x</td>
<td class="org-left"> #1358</td>
<td class="org-left">&#xa0;</td>
<td class="org-left">x</td>
</tr>

<tr>
<td class="org-left">t</sub></td>
<td class="org-left">&#xa0;</td>
<td class="org-left">x</td>
<td class="org-left">&#xa0; </td>
<td class="org-left">x</td>
</tr>

<tr>
<td class="org-left">triangular</td>
<td class="org-left">x</td>
<td class="org-left">x</td>
<td class="org-left">x</td>
<td class="org-left">#1328</td>
</tr>

<tr>
<td class="org-left">uniform</td>
<td class="org-left">x</td>
<td class="org-left">x</td>
<td class="org-left">x</td>
<td class="org-left">x</td>
</tr>

<tr>
<td class="org-left">vonmises</td>
<td class="org-left">x</td>
<td class="org-left">x</td>
<td class="org-left">x</td>
<td class="org-left">#1330</td>
</tr>

<tr>
<td class="org-left">wald</td>
<td class="org-left">x</td>
<td class="org-left">x</td>
<td class="org-left">x</td>
<td class="org-left">x</td>
</tr>

<tr>
<td class="org-left">weibull</td>
<td class="org-left">x</td>
<td class="org-left">x</td>
<td class="org-left">x</td>
<td class="org-left">x</td>
</tr>

<tr>
<td class="org-left">zipf</td>
<td class="org-left">x</td>
<td class="org-left">#1350</td>
<td class="org-left">&#xa0;</td>
<td class="org-left">&#xa0;</td>
</tr>

</tbody>
</table>

-   The list of distributions available in Numpy was retrieved from the [documentation](https://numpy.org/doc/stable/reference/random/generator.html#numpy.random.Generator).
-   Aesara transpiles RVs to Numba following the rules in [this file](https://github.com/aesara-devs/aesara/blob/main/aesara/link/numba/dispatch/random.py).
-   Aesara automatically transpiles RVs with name `name` to `jax.random.name`. List of available distributions in the [JAX documentation](https://jax.readthedocs.io/en/latest/jax.random.html).

Should we open issues for distributions that are missing from `numpy` / for which there is no Numba implementation?

## A separate repo for probability distributions?

There are also many distributions that are implemented e.g. in scipy and that can be useful for bayesian inference (Wishart for instance). We probably don&rsquo;t want to bloat `Aesara` with too many distributions, then I&rsquo;d like to start a discussion about opening a separate repo (`aerand`?) with distributions that are frequently used in bayesian inference. This would have several advantages:

1.  Keep the number distributions in `Aesara` to a minimum;
2.  We could remove distribution-specific logic out of `aeppl` by moving the `<name>_logprob` and `<name>_logcd` funtions to this repo;
3.  We could add extra information about random variables, such as their distribution&rsquo;s support and constraints on their parameters (see [MCX for an example of constraints implementation](https://github.com/rlouf/mcx/blob/master/mcx/distributions/constraints.py)). This information can be used to remove distribution-specific logic for the transforms in `aeppl`, but also for static validation. For instance `aeppl` could return a warning when asked to build the logprob graph for a model where parameters have incompatible support. For instance a `Normal` prior on a standard deviation.

**Comment by brandonwillard** [comment]
> A separate repo for probability distributions?

I've often thought that we should have a separate repo for SciPy `Op`s in general.

**Comment by ricardoV94** [comment]
We could implement some Numba / Jax distributions that don't have an immediate Op in a backend by using a more clever dispatch (or even a Aesara rewrite). For instance in JAX: `Binomial(n, p) = sum(Bernoulli(p, size=n))` (they really don't have a binomial?)

**Reply by rlouf** [reply]
No they have very few probability distributions :/
